src/mesh/forest/forest.hpp

```diff
@@ -0,0 +1,151 @@
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef MESH_FOREST_FOREST_HPP_
+#define MESH_FOREST_FOREST_HPP_
+
+#include <array>
+#include <map>
+#include <memory>
+#include <set>
+#include <tuple>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "basic_types.hpp"
+#include "defs.hpp"
+#include "mesh/forest/tree.hpp"
+#include "mesh/logical_location.hpp"
+#include "utils/bit_hacks.hpp"
+#include "utils/indexer.hpp"
+
+namespace parthenon {
+namespace forest {
+
+class Forest {
+  bool gids_resolved = false;
+  std::unordered_map<std::int64_t, std::shared_ptr<Tree>> trees;
+
+ public:
+  int root_level;
+  int forest_level;
+
+  void AddTree(const std::shared_ptr<Tree> &in) {
+    if (trees.count(in->GetId())) {
+      PARTHENON_WARN("Adding tree to forest twice.");
+    }
+    trees[in->GetId()] = in;
+  }
+
+  int AddMeshBlock(const LogicalLocation &loc, bool enforce_proper_nesting = true) {
+    gids_resolved = false;
+    return trees[loc.tree()]->AddMeshBlock(loc, enforce_proper_nesting);
+  }
+  int Refine(const LogicalLocation &loc, bool enforce_proper_nesting = true) {
+    gids_resolved = false;
+    return trees[loc.tree()]->Refine(loc, enforce_proper_nesting);
+  }
+  int Derefine(const LogicalLocation &loc, bool enforce_proper_nesting = true) {
+    gids_resolved = false;
+    return trees[loc.tree()]->Derefine(loc, enforce_proper_nesting);
+  }
+
+  std::vector<LogicalLocation> GetMeshBlockListAndResolveGids();
+
+  int count(const LogicalLocation &loc) const {
+    if (trees.count(loc.tree()) > 0) {
+      return trees.at(loc.tree())->count(loc);
+    }
+    return 0;
+  }
+
+  RegionSize GetBlockDomain(const LogicalLocation &loc) const {
+    return trees.at(loc.tree())->GetBlockDomain(loc);
+  }
+
+  std::array<BoundaryFlag, BOUNDARY_NFACES>
+  GetBlockBCs(const LogicalLocation &loc) const {
+    return trees.at(loc.tree())->GetBlockBCs(loc);
+  }
+
+  std::vector<NeighborLocation> FindNeighbors(const LogicalLocation &loc, int ox1,
+                                              int ox2, int ox3) const {
+    return trees.at(loc.tree())->FindNeighbors(loc, ox1, ox2, ox3);
+  }
+
+  std::vector<NeighborLocation> FindNeighbors(const LogicalLocation &loc) const {
+    return trees.at(loc.tree())->FindNeighbors(loc);
+  }
+  std::size_t CountMeshBlock() const {
+    std::size_t count{0};
+    for (auto &[id, tree] : trees)
+      count += tree->CountMeshBlock();
+    return count;
+  }
+
+  // TODO(LFR): Probably eventually remove this. This is only meaningful for simply
+  // oriented grids
+  LogicalLocation GetAthenaCompositeLocation(const LogicalLocation &loc) const {
```

Comment:
I thought we came up with some name we liked better than Athena? UniTree?

Reply:
Yggdrasil?

Reply:
lol. I'm in.